### PR TITLE
[Frontend, MLIR] Support indexing of the dynamically shaped arrays

### PR DESCRIPTION
**Pull request by grwlf**
In this PR we enable indexing for tensors with dynamic shapes. This PR is intended to be merged after the #370 

* [x] Indexing with constant
* [x] Indexing with variable
* [x] Modifications 

[sc-47632]

The corresponding [Jax PR](https://github.com/google/jax/pull/19083) suggests the fix to the upstream.

### CODE REVIEW

**Comment by grwlf** [comment]
Hi @rmoyard , could you please review the MLIR-related parts of this PR?

File: mlir/lib/Catalyst/Transforms/ScatterPatterns.cpp [hunk 1]
```diff
@@ -366,7 +366,7 @@ struct ScatterOpRewritePattern : public mlir::OpRewritePattern<mhlo::ScatterOp>
                     auto indexScatter =
                         builder.create<tensor::ExtractOp>(loc, scatterIndices, index);
                     auto indexUpdateCasted =
-                        builder.create<index::CastSOp>(loc, builder.getI32Type(), indexUpdate);
+                        builder.create<index::CastSOp>(loc, indexScatter.getType(), indexUpdate);
```
Comment (rmoyard) [inline mlir/lib/Catalyst/Transforms/ScatterPatterns.cpp:369]: Thanks, good catch :+1:

File: frontend/catalyst/utils/jax_extras.py [hunk 2]
```diff
@@ -994,3 +1018,125 @@ def bind(self, *args, **params):
...
+    dimension_numbers,
+    slice_sizes,
+    unique_indices,
+    indices_are_sorted,
+    mode,
+    fill_value,
+):  # pragma: no cover
+    """This is a modified version of the ``jax._src.lax.slicing._gather_shape_rule`` function. In
+    this version we skip the dimension check if the dimenstions are dynamic. The same fix is
+    suggested to the upstream as https://github.com/google/jax/pull/19083"""
```
Comment (dime10) [inline frontend/catalyst/utils/jax_extras.py:1036]: We should still include their license & copyright notice here, even if this function is copied temporarily. See example here: 
https://github.com/PennyLaneAI/catalyst/blob/5e177244e68b6eac3c4a23747ab433f9821c3134/frontend/catalyst/utils/jax_extras.py#L367-L371

**Review by dime10 — APPROVED** [review]
Nice patch 👍

File: frontend/catalyst/utils/jax_extras.py (hunk 2 above)
Comment (josh146) [reply frontend/catalyst/utils/jax_extras.py:1036]: 👍

**Comment by grwlf** [comment]
@josh146 I meant this PR, when I asked about CodeFactor problems.

File: frontend/catalyst/utils/jax_extras.py [hunk 3]
```diff
@@ -463,10 +474,23 @@ def abstractify(args, kwargs):
...
+    )
+    register_lowering(gather2_p, _gather_lower)
+
     @wraps(fun)
     def make_jaxpr_f(*args, **kwargs):
         # TODO: re-use `deduce_avals` here.
-        with Patcher((jax._src.interpreters.partial_eval, "get_aval", get_aval2)), ExitStack():
+        with Patcher(
+            (jax._src.interpreters.partial_eval, "get_aval", get_aval2),
+            (jax._src.lax.slicing, "gather_p", gather2_p),
```
Comment (rmoyard) [inline frontend/catalyst/utils/jax_extras.py:492]: @dime10 @grwlf here we have an issue, the new rule does not define jvp and therefore it is not compatible with grad or jacobian transformations ad.defjvp(gather_p, _gather_jvp_rule, None) see the original gather_p

```
gather_p = standard_primitive(
    _gather_shape_rule, _gather_dtype_rule, 'gather',
    weak_type_rule=_argnum_weak_type(0))
ad.defjvp(gather_p, _gather_jvp_rule, None)
ad.primitive_transposes[gather_p] = _gather_transpose_rule
batching.primitive_batchers[gather_p] = _gather_batching_rule
pe.padding_rules[gather_p] = _gather_pad_rule
```

Comment (josh146) [reply frontend/catalyst/utils/jax_extras.py:492]: @rmoyard does this result in a user-facing bug?

Comment (rmoyard) [reply frontend/catalyst/utils/jax_extras.py:492]: Yes you cannot use jax.grad inside qjit when a gather operation is created, for example slicing.

Comment (josh146) [reply frontend/catalyst/utils/jax_extras.py:492]: Ah, so #305?

Comment (dime10) [reply frontend/catalyst/utils/jax_extras.py:492]: That issue is about Catalyst gradients, I think Romain is talking about JAX gradients (which are run in the frontend on the jaxpr, hence the primitives need gradient rules).

Comment (josh146) [reply frontend/catalyst/utils/jax_extras.py:492]: Oh, got it!

Comment (rmoyard) [reply frontend/catalyst/utils/jax_extras.py:492]: Yes exactly what David said, qjit with jax.grad and slicing

Comment (rmoyard) [reply frontend/catalyst/utils/jax_extras.py:492]:
```
qjit(jax.grad(f))(x)
```

Comment (josh146) [reply frontend/catalyst/utils/jax_extras.py:492]: Just checking if we have a resolution on this particular comment thread :)

Comment (dime10) [reply frontend/catalyst/utils/jax_extras.py:492]: @grwlf How is the upstream PR looking for this patch? 
In the meantime, can we just attach the original gradient rule to the patched primitive?
